Support block-specific style overrides

hmedina opened the pull request:
## Purpose

Allow local overrides of the Pygments style, on a per-code-block basis.

This work extends the directives for `code-block`, `sourcecode`, `literalinclude` and `code`, by introducing two new options, `style-light` and `style-dark`. As the app parses the document and discovers these, the builders collects the information. This allows Pygments to produce CSS or STY files (for HTML & LaTeX) for the relevant classes. These are moved from the initializing phase of the builder, to the finalizing phase.

Special attention was paid to light & dark modes. For most builders (e.g. LaTeX), this distinction is ignored, with only the `light` style having any effect; this mirrors how currently, the configuration value `pygments_dark_style` has no effect on such builders. For the HTML family builders, if the theme supports a dark style, then the `pygments_dark.css` file gains the appropriate selectors, independently of what happens to the `pygments.css` file for light styles. Since in the general sense, a user need not track the overrides for light style in the same fashion as for the dark style, the selectors are generated by hashing the docutils node. In a CPython implementation, this returns a pointer; as the docutils hierarchy is a tree, a pointer to the code-block appears sufficient for these purposes, as collisions should not happen; this said, I'm open to suggestions. The single-style builders, like LaTeX, use a simpler tracking for their command prefixes (used similarly as CSS selectors in the STY files).

The files have been linted with Ruff.

The new options are tested in the test suite. I'm unsure how to incorporate more advanced testing for other parts, so I welcome feedback (should I add tests to the builders?).


## References

- Discussion (with examples!) : https://github.com/orgs/sphinx-doc/discussions/13524
- https://github.com/sphinx-doc/sphinx/issues/9105#issuecomment-821844520

hmedina committed:
Working protype!

* Code highlight directives accept two new options, style-light and style-dark; these override the default Pygments style with a block-specific one, but only for the block in question
* Classes amended were CodeBlock, LiteralInclude, and Code; this means directives code-block, sourcecode, literalinclude, and code
* A theme may support light and/or dark styles; as already done, light and dark styles must be tracked separately. For this overhaul, a data structure is created inside the HtmlBuilder; this tracks the relevant styles, their associated PygmentsBridge highlighter, and the code blocks they're associated with.
* Association with code blocks is handled by using the docutil's node hash; this is used as a CSS selector.
* As the Html5Translator visits elements of the docutils tree, it discovers their light and dark styles, if any. It populates the HtmlBuilder's data structure.
* To finalize, the CSS sheets are created. Since the Pygments style is now a "generated" file rather than static (it depends on what style overrides document contains), the function call is moved down: the HtmlBuilder's create_pygments_style_file got moved from copy_static_files() into a task under the finish() method.
* CSS selectors are grouped to supply different code blocks with the same override style, if the user so chooses (Pygments handles this internally if presented the selectors). This minimizes the length of the CSS files. They're also annotated with a comment for their override style.

hmedina committed:
Media queries for light or dark style

* There is a CSS media feature that is set by the user agent: https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme ; we add those specifiers to the HTML builder's light and dark Pygments style sheet writers

hmedina committed:
Merge branch 'sphinx-doc:master' into master

hmedina committed:
Merge branch 'sphinx-doc:master' into master

hmedina committed:
Unify pygments.css and pygments_dark.css

* Use media queries and scopes to include light and dark styles in a single CSS file
* `app.registry.css_files` apparently only tracked "extra" files, but not the default `pygments.css` file, so the `test_theming.test_dark_style` block removed that check
* Added a `stylename` attribute to `PygmentsBridge` objects, so they can track the "pretty" style name used throughout Pygments, rather than relying on the classname of the associated style, which is not necessarily what is declared in the Pygments style entry-point installation

hmedina committed:
Revert "Unify pygments.css and pygments_dark.css"

This reverts commit b5c34d21dcec89d0e1caaef79300f952fddf95ab.

hmedina committed:
Use separate style sheets; leave media queries out

* `python_docs_theme` uses separate CSS files for light and dark modes, with a simple javascript function to toggle what is shown. It doesn't decorate either CSS file with a `@media` query. This seems advantageous, allowing the user to select their preferred view, and aligns with the philosophy that "the theme is not part of the document, but just a view of it" (paraphrased from MDN)
* bugfix: specialized dark iterator used wrong value for CSS sheet building

hmedina committed:
Fix `singlehtml` builder

hmedina committed:
Remove unnecessary parenthesis

hmedina committed:
Add LaTeX builder

* As for the HTML builder, the building of the style sheet is moved from the `prepare_writing` method to the `finish` one; overriding styles are discovered on-the-fly, so the `.sty` file can only be finalized after the document's nodes have been visited

hmedina committed:
Merge branch 'sphinx-doc:master' into master

hmedina committed:
Lint & style with Ruff

hmedina committed:
Provide usage examples in docs for new directive options

hmedina committed:
Merge branch 'sphinx-doc:master' into master

hmedina committed:
Add name to contributor list

hmedina committed:
Merge branch 'master' into master

hmedina committed:
Sphinx-lint; remove trailing whitespaces

hmedina committed:
Bugfix: fetcher of PygmentsBridge objects from string had wrong parameter

hmedina committed:
Properly catch Nones; pass `mypy`

hmedina committed:
Ruff, round 2

hmedina committed:
Ruff, formatting

hmedina committed:
Fix MyPy ?

hmedina committed:
Merge remote-tracking branch 'upstream/master'

hmedina committed:
Fix docs-lint; line too long after adding PR number

jfbu commented on sphinx/builders/latex/__init__.py:
```diff
@@ -275,6 +276,23 @@ def init_multilingual(self) -> None:
+        """Add a styler to the tracker of highlighting styles."""
+        if style not in self.specialized_highlighters:
+            pb = highlighting.PygmentsBridge(dest='latex', stylename=style)
+            pb.formatter_args['commandprefix'] = 'PYG' + style
```
Quite a few Pygments style names contain a dash or an underscore (`pygmentize -L styles`; I vaguely remember one with a digit) which can not serve in LaTeX macro names.  The definitions ending up in `sphinxhighlighting.sty` will use `\@namedef` but the actual output mark-up in the produced `project.tex` file will break on latex compilation.

I think the https://github.com/gpoore/minted LaTeX package had to solve that problem as it was using in the past about same prefix formation as the one here.

jfbu reviewed (commented):
warning: this got a bit long.

In sphinx.highlighting you can see that `_LATEX_ADD_STYLES`  appends some definitions of its own with `PYG` prefix. They are involved in sphinx/texinputs/sphinxlatexliterals.sty.  They are used for linebreaks of inline literals, and here that should be fine, as I suspect your PR does not modify them.  But the support for the `verbartimwrapslines` and `verbartimforcewraps` option of `'sphinxsetup'` (search for them at https://www.sphinx-doc.org/en/master/latex.html) crucially depends on the fact that the `\PYG` macro is used in the mark-up.  So the wrapping of long code lines will fail for the specially highlighted code-blocks.

The first thing which comes to mind is that the mark-up should use `\PYG[stylename]` and a quite non trivial change inside sphinxlatexliterals.sty around `\def\spx@verb@wrapPYG@i` and similar macros. (my thoughts evolved while writing this see below)

I think it is best to delay that to a later stage which we can take care of, but it will be impossible with the current choice of `\PYGstylename` command.  Unfortunately my immediate thinking is that solving this requires a Pygments pull request that they output `\PYG[stylename]` mark-up optionally, and not only `\PYG`.  The redefinition of `\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}` so that `\PYG` becomes a macro testing for optional argumant `[..]` can be achieved by ourselves and put at end of `sphinxhighlighting.py`, and the other material there can still use the stuff such as `\@namedef{PYGautumn@tok@w}` from the modified command prefix, taht is not a problem.  The essential point is that Pygments output LaTeX mark-up optionally uses not the command prefix but `\PYG[something]`.  By the way this would solve the problem with style names containing dashes, underscores or digits.

No wait, we could use the trick of `pb.formatter_args['commandprefix'] = 'PYG[' + style + ']'` (*edited*) (or better `f'PYG[{style}]'`) and fix ourself the stylesheets returned by Pygments (the `\@namedef` parts will need no change, the other parts now will make many various definitions of `\PYG` which we will have to fix).  Unfortunately this will add one more layer of LaTeX, and future Sphinx maintainers will have a hard time maintaining because it requires too much TeX knowledge.  However at this stage I see no better way.  There is another way to define dynamically the core TeX code one finds in sphinxlatexliterals.sty and to activate it via a hook into sphinxVerbatim.  This looks even more terrible from a maintenance point of view.

Maybe at this stage we would only need some simple work around for the problem of hyphens and dashes, maybe simply repalce all such characters by a `Z`.  At a later stage, the poor Sphinx LaTeX maintainer will have to implement the `[...]` method.  So contrarily to what I initially thought, I may not have at this stage to require changes, but I have. only tested on basic things.

As per the handling of dark-mode, it could be possible to do it, I have actually done it for a project using Sphinx, this requires modifying the Verbatim background color which can be done by `\sphinxsetup` anywhere inside the document, something such as (with the choices I made for the colors)
```
\sphinxsetup{%
    pre_background-TeXcolor={RGB}{40,42,54},% #282a36 aka VerbatimColor
    pre_TeXcolor={RGB}{248,248,242},% #f8f8f2
    VerbatimHighlightColor=VerbatimColor!75,
}
```
but then it has to be reset later.

I also needed to add `pygments_style = pygments_dark_style` in the conf.py.

jfbu reviewed (changes requested):
So to recapitulate:
- as it stands the PDF build will break if the Pygments stylename contains dash, hyphen or digit,
- one can simply replace problematic characters with say a `Z` and then, basically the whole thing works for PDF via LaTeX
- but the wrapping of long codelines will work less well, because `\sphinxbreaksatspecials` requires the `\PYGZob` etc... definitions, so breaking will only happens at spaces,
- and the `verbatimforcewraps` option of `'sphinxsetup'` will be completely without effect.

We have considered a method [^1] [^2] to circumvent this, but it would not be reasonable for it to be part of this PR.  So the PR has to fix the issues with style names, even though maybe in future to re-establish wrapping of long code lines we will have to solve it in another manner.

[^1]: I apologize but I underestimated the difficulty with using `f'PYG[{style}]'` as command prefix. This can not work. The Pygments stylesheet does not use `\@nameuse` inside its `\@namedef` definitions, and for some TeXnical reasons it thus is impossible to give `\PYG` a suitable definition as I had envisioned.  I see an alternative using `f'PYG_{style}_'` but I shall not give too many TeX/LaTeX details here as this PR should not include such complex LaTeXnicalities.

[^2]: sadly `pygmentize -f latex -P 'commandprefix=PYG_autumn_' temp.py` with `temp.py` solely containing `r"\"` produces broken output (although the stylesheet is ok). So no, one can not use `_` in the style name (too bad as Sphinx would have further escaped it to `\_` which is useful for TeX delimited macro).

hmedina committed:
"sanitize" LaTeX command prefix from user-specified style

* Whether a character is in the correct category for the definition of a LaTeX macro can, in the general sense, only be determined at compile time. As a fall-back, "valid" macro names generally use the a-zA-Z range; so we use this to replace any character not in those ranges with an uppercase Z using regular expressions.
* Since the user-facing names come from the SetupTools installation entry-points, they need not match the `name` attribute in the associated classes (sigh). Added to this, something needs to be printed to the LaTeX file. So, on the Python side and any user-facing usage retain the user-given name (this also helps avoid collisions, as Z-replacement loses information); any LaTeX-printed file will use the "sanitized" version in the macro names where likely-wrong-category characters got replaced with a Z. A header / comment in the .sty file specifies the user-given name for that portion of the file
* The sphinx re-definitions of "problematic" LaTeX characters (e.g. \PYGam{\&} ) gets overhauled; the string now contains an `override` key, for use with Python's `str.format(key=value)` syntax. When given an empty string, the string prior to this commit is returned; otherwise, appropriate overrides are generated for the various style-specific special characters. However, the usage in `sphinx/texinputs/sphinxlatexliterals.sty`, lines 561-650, seem to have the `PYG` prefix baked-in...

hmedina committed:
Fix ruff format?

jfbu commented on sphinx/highlighting.py:
```diff
@@ -229,12 +230,36 @@ def highlight_block(
+                    logger.error(
+                        __('Encountered %s in selectors field; expected a string for the '
+                           'LaTeX formatter, using default values as fallback.\n'
+                           'Please report his error.'),
```
typo

jfbu commented on sphinx/builders/latex/__init__.py:
```diff
@@ -280,7 +281,7 @@ def add_block_style(self, style: str) -> None:
         if style not in self.specialized_highlighters:
             pb = highlighting.PygmentsBridge(dest='latex', stylename=style)
-            pb.formatter_args['commandprefix'] = 'PYG' + style
+            pb.formatter_args['commandprefix'] = 'PYG' + re.sub(r'[^a-zA-Z]', 'Z', style)
```
due to `Z` showing up already in the default mark-up such as `\PYGZbs` I would prefer the replacement letter here to be say `Y` which will be easier to handle in the TeX hacks needed to apply "verbatimforcewraps" hard line-breaks

jfbu commented on sphinx/highlighting.py:
```diff
@@ -229,9 +230,40 @@ def highlight_block(
+                        __(
+                            'Encountered %s in selectors field; expected a string for the '
+                            'LaTeX formatter, using default values as fallback.\n'
+                            'Please report his error.'
```
I think there is a typo s/his/this/ here

jfbu commented on sphinx/highlighting.py:
```diff
@@ -73,24 +74,24 @@
+\def\PYG{override}Zhy{{\sphinxhyphen}}% defined in sphinxlatexstyletext.sty
+\def\PYG{override}Zsq{{\text\textquotesingle}}
+\def\PYG{override}Zdq{{"}}
+\def\PYG{override}Zti{{\text\textasciitilde}}
```
This is no good due to internals of `sphinxlatexliterals.sty` (see  aournd `\newcommand*\sphinxbreaksbeforelist`).  What the code should do if override is not `''` is to do
```
\def\PYG{override}Zbs{\PYGZbs}
```
etc...

See my other comment about `_LATEX_ADD_STYLES` from highlighting.py.

jfbu commented on sphinx/highlighting.py:
```diff
@@ -229,9 +230,40 @@ def highlight_block(
+                stylesheet = self.formatter(
+                    commandprefix='PYG' + tex_name
+                ).get_style_defs()
+                sphinx_redefs = _LATEX_ADD_STYLES.format(override=tex_name)
```
Due to the way this is separated from other (non override) branch we could use another constant `_LATEX_ADD_EXTRA_STYLES` here which would make easier implementing my remark about `\PYGZbs` et al.

jfbu commented on sphinx/highlighting.py:
```diff
@@ -229,9 +230,40 @@ def highlight_block(
-            return formatter.get_style_defs() + _LATEX_ADD_STYLES
+            if selectors:
+                if isinstance(selectors, str):
+                    tex_name = re.sub(r'[^a-zA-Z]', 'Z', selectors)
```
I would prefer `Y` letter here because Pygments has a hardcoded `Z` in output for special character macro escapes such as `\PYGZbs`.  Having a `Y` here could help the TeX macro layer.

jfbu replied:
There is an issue with `style_1`, `style_2` etc.. which are all mapping to `styleYY`.

We could drop the idea of the `Z` or `Y` and use rather something such as `hashlib.md5(style.encode()).hexdigest()[:8]` followed by replacement of `0` to `9` by `G` to `P`.  I limit here to 8 letters the produced sanitized version to avoid the output to be too voluminous, perhaps we can take the risk of only 6 letters.   It is sometimes needed to manually investigate Sphinx LaTeX output and it should not have too gigantic TeX macro names.

The original style name is in the stylesheet.

jfbu commented on sphinx/builders/latex/__init__.py:
```diff
@@ -275,6 +277,25 @@ def init_multilingual(self) -> None:
+            pb = highlighting.PygmentsBridge(dest='latex', stylename=style)
+            pb.formatter_args['commandprefix'] = 'PYG' + re.sub(
+                r'[^a-zA-Z]', 'Z', style
+            )
```
Sorry for one more remark.  As I have argued in previous review, ultimately we will have to communicate the sanitized stylename to the sphinxVerbatim environment.  But then we *do not need* the markup to use any other mark-up than `\PYG`.
1. the special escaped characters `\PYGZbs` etc... have anyhow to be defined identically or expand to the default one for reasons argued elsewhere, the simplest is for them to use the exact same mark-up. in all styles,
2. the `\PYG` command will now be attributed a special meaning [see below] which will use a definition using some macro `\sphinxpygmentsstyletexname` or the like.  It will then dispatch to the macros from the stylesheet.
3. the stylesheet macros will all have been defined with the special (sanitized) command prefix.

In brief this formatter has to be used with two distinct command prefixes: the standard one `PYG` for producing the highlighting mark-up ending up inside the sphinxVerbatim environment, and the other one `'PYG' + sanitized_name` for the style sheets.

The `_LATEX_ADD_STYLES` will be needed *only* for the default style.  For the custom block styles it should *not* be added to the Pygments stylesheet output.  And at end of this `_LATEX_ADD_STYLES` we should now have something such as this
```tex
\protected\def\PYG{\ifdefined\sphinxpygmentsstyletexname
    \csname PYG\sphinxpygmentsstyletexname\expandafter\endcsname
  \else
    \expandafter\spx@PYG
  \fi
}
\def\spx@PYG#1#2{\PYG@reset\PYG@toks#1+\relax+{\PYG@do{#2}}}
```
Again, the output mark-up found inside the produced `.tex` file should always use `\PYG` macro (ie `'PYG'` command prefix).  But the stylesheets with use `'PYG' + sanitizedtexname` and inside the `sphinxVerbatim` environment for custom styled blocks and *only them* we will have an extra `\def\sphinxpygmentsstyletexname{...}` line where the dots stand for the sanitized name discussed elsewhere, using only letters from A to P.

All `sphinxVerbatim` environments will use the `\PYG` command but its definition is altered as above to dispatch to the commands defined in the stylesheets, which are all accumulated inside `sphinxhighlighting.sty` in the LaTeX build directory.

jfbu committed:
LaTeX: fix style arg lacking, refactor to achieve full LaTeX support

jfbu committed:
Merge branch 'master' into multistyle

jfbu reviewed (approved):
Regarding the fact that this works fully for LaTeX/PDF output, LGTM!  Internal aspects may need reviewing though.

jfbu committed:
Remove left-over curly brace pairs in LaTeX code in highlighting.py

hmedina committed:
Merge branch 'sphinx-doc:master' into master

hmedina committed:
Simplify LaTeX code

* `LaTeXBuilder.get_bridge_for_style()` was only used to get a recently-created `PygmentsBridge` object, but added a None-check. Refactoring the method that added the created object to return it (i.e. `LaTeXBuilder.add_block_style()`) avoids the extra method and the requirement for a None check. If the method fails, it would fail at creation of the PygmentsBridge object, which is better handled by its own reporting
* Rename `LaTeXBuilder.add_block_style()` to `update_override_styles()`

hmedina committed:
Simplify HTML code

* `StandaloneHTMLBuilder.get_bridge_for_style()` was only used to get a recently-created `PygmentsBridge` object, but added a None-check. Refactoring the methods that added the created object to return it (i.e. `add_block_dark_style()` & `add_block_light_style()`) avoids the extra method and the requirement for a None check. If the methods fail, they would fail at creation of the PygmentsBridge object(s), which is better handled by its own reporting
* Rename `add_block_dark_style()` to `update_override_styles_dark()`
* Rename `add_block_light_style()` to `update_override_styles_light()`

jfbu committed:
LaTeX: implement support for custom background color

Some styles set the background_color to #ffffff, in such cases we
ignore that because the default Sphinx PDF light gray is nicer.

This also handles a "default" text color, but some testing shoud be
done.
